channels/caches.go

```diff
@@ -0,0 +1,133 @@
+package channels
+
+import (
+	"sync"
+	"sync/atomic"
+
+	datatransfer "github.com/filecoin-project/go-data-transfer"
+)
+
+type readIndexFn func(datatransfer.ChannelID) (int64, error)
+
+type cacheKey struct {
+	evt  datatransfer.EventCode
+	chid datatransfer.ChannelID
+}
+
+type blockIndexCache struct {
+	lk     sync.RWMutex
+	values map[cacheKey]*int64
```

Comment:
out of interest: how do we manage blocks as _indexes_, is that something we defer to the protocol and we just record a count? if we can have duplicate indexes, is that something we care about here or does that somehow bubble up from the protocol?

Reply:
we record a count from the protocol. honestly I'm a bit ambiguous about indexes in general filtering up to data transfer but I think for now I prefer it.